elyra/pipeline/component_registry.py

```diff
@@ -0,0 +1,156 @@
+
+
+class ComponentReader(SingletonConfigurable):
+    _type = 'local'
+
+    def read_component_definition(self, location: str):
+        raise NotImplementedError()
+
+
+class FilesystemComponentReader(ComponentReader):
+    _type = 'filename'
+    _dir_path: str = ''
+
+    def read_component_definition(self, location):
+        component_file = os.path.join(os.path.dirname(__file__), location)
+
+        if not os.path.exists(component_file):
+            raise FileNotFoundError(f'Invalid component location: {location}')
+
+        with open(component_file, 'r') as f:
+            return f.readlines()
+
+
+class UrlComponentReader(ComponentReader):
+    _type = 'url'
+    _url_path: str
+
+    def read_component_definition(self, location):
+        res = requests.get(location)
+        if res.status_code != HTTPStatus.OK:
+            raise FileNotFoundError(f'Invalid component location: {location}')
+
+        return res.text
+
+
+class ComponentRegistry(LoggingConfigurable):
+    readers = {
+        FilesystemComponentReader._type: FilesystemComponentReader(),
+        UrlComponentReader._type: UrlComponentReader()
+    }
+
+    def __init__(self, component_catalog_location: str, parser: ComponentParser):
+        super().__init__()
+        self._component_catalog_location = component_catalog_location
+        self._parser = parser
+        self.log.info(f'Creating new registry using {self.catalog_location}')
+
+    @property
+    def catalog_location(self) -> str:
+        return self._component_catalog_location
+
+    def get_all_components(self):
+        """
+        Builds a component palette in the form of a dictionary of components.
+        """
+
+        print(f'Retrieving components from {self.catalog_location}')
+
+        # Get components common to all runtimes
+        components = {}
+        components['categories'] = list()
+
+        # Process component catalog
+        component_catalog = self._read_component_catalog()
+
+        # process each registered component
+        for component_entry in component_catalog:
+            self.log.debug(f"Component registry found component {component_entry['name']}")
+
+            # Get appropriate reader in order to read component definition
+            reader = self._get_reader(component_entry)
+
+            component_definition = \
+                reader.read_component_definition(component_entry['location'][reader._type])
+
+            # Parse the component definition in order to add to palette
+            component_json = self._parser.parse_component_details(component_definition, component_entry['name'])
+            if component_json:
+                components['categories'].append(component_json)
+
+        return components
+
+    def get_properties(self, processor_type, component_id):
+        """
+        Return the properties JSON for a given component.
+        """
+
+        # Process component catalog
+        component_catalog = self._read_component_catalog()
```

Comment:
Perhaps we should make `component_catalog` and singleton object: `ComponentCatalog` who's init is `_read_component_catalog()`.  This would save duplicate reads here and in the above method - `get_all_components()`.

`ComponentCatalog` would provide single component get() method as well as a coroutine to get all the components.

`ComponentCatalog.get_component(id: str)` would return a `Component` instance and raise a `ValueError` if no component was found - just as below.  The `Component` instance would then know its reader and could provide 'location', 'body' and 'properties' properties.

Looks like `ComponentRegistry` could be easily converted to provide all of this functionality.